exercises/src/main/scala/fpinscala/state/State.scala

```diff
@@ -30,34 +30,118 @@ object RNG {
+    val (d1, r1) = double(rng)
+    val (d2, r2) = double(r1)
+    val (d3, r3) = double(r2)
+    ((d1, d2, d3), r3)
+  }
+
+  // 6.4 list of random Ints
+  def ints(count: Int)(rng: RNG): (List[Int], RNG) = {
+    if (count <= 0) (Nil, rng)
+    else {
+      val (i, r1) = rng.nextInt
+      val (l, r2) = ints(count - 1)(r1)
+      (i :: l, r2)
+    }
+  }
 
-  def double(rng: RNG): (Double, RNG) = ???
+  // 6.5 use map to implement double
+  def doubleMap: Rand[Double] = map(nonNegativeInt)(i => i / (Int.MaxValue.toDouble + 1))
 
-  def intDouble(rng: RNG): ((Int,Double), RNG) = ???
+  // 6.6
+  def map2[A,B,C](ra: Rand[A], rb: Rand[B])(f: (A, B) => C): Rand[C] = rng => {
+    val (a, r2) = ra(rng)
+    val (b, r3) = rb(r2)
+    (f(a, b), r3)
+  }
+
+  // 6.7
+  def sequence[A](fs: List[Rand[A]]): Rand[List[A]] = {
+    fs.foldRight(unit(List[A]()))((ra, rb) => map2(ra, rb)(_ :: _))
+  }
 
-  def doubleInt(rng: RNG): ((Double,Int), RNG) = ???
+  // 6.7 use sequence and List.fill(n)(x) to implement ints
+  def ints2(count: Int): Rand[List[Int]] = {
+    sequence(List.fill(count)(int))
+  }
 
-  def double3(rng: RNG): ((Double,Double,Double), RNG) = ???
+  // 6.8
+  def flatMap[A,B](f: Rand[A])(g: A => Rand[B]): Rand[B] = rng => {
+    val (a, r) = f(rng)
+    g(a)(r)
+  }
 
-  def ints(count: Int)(rng: RNG): (List[Int], RNG) = ???
+  // 6.8 use flatMap to implement nonNegativeLessThan
+  def nonNegativeLessThan(n: Int): Rand[Int] = flatMap(nonNegativeInt) {
+    i =>
+      val mod = i % n
+      if (i + (n - 1) - mod >= 0) unit(mod) else nonNegativeLessThan(n)
+  }
 
-  def map2[A,B,C](ra: Rand[A], rb: Rand[B])(f: (A, B) => C): Rand[C] = ???
+  // 6.9 reimplement map and map2 in terms of flatmap
+  def mapfm[A,B](s: Rand[A])(f: A => B): Rand[B] = {
+    flatMap(s)(a => unit(f(a)))
+  }
 
-  def sequence[A](fs: List[Rand[A]]): Rand[List[A]] = ???
+  def map2fm[A,B,C](ra: Rand[A], rb: Rand[B])(f: (A, B) => C): Rand[C] = {
+    flatMap(ra)(a => map(rb)(b => f(a, b)))
+  }
 
-  def flatMap[A,B](f: Rand[A])(g: A => Rand[B]): Rand[B] = ???
 }
 
+// 6.10 Generalize the functions unit, map, map2, flatMap, and sequence
+// on the State case class where possible, otherwise on the companion object
+
 case class State[S,+A](run: S => (A, S)) {
-  def map[B](f: A => B): State[S, B] =
-    ???
-  def map2[B,C](sb: State[S, B])(f: (A, B) => C): State[S, C] =
-    ???
-  def flatMap[B](f: A => State[S, B]): State[S, B] =
-    ???
+  def map[B](f: A => B): State[S, B] = flatMap(a => unit(f(a)))
+
+  def map2[B,C](sb: State[S, B])(f: (A, B) => C): State[S, C] = flatMap(a => sb.map(b => f(a, b)))
+
+  def flatMap[B](f: A => State[S, B]): State[S, B] = State(s1 => {
+    val (a, s2) = run(s1)
+    f(a).run(s2)
+  })
+
+  def sequence[S,A](fs: List[State[S,A]]): State[S,List[A]] = {
+    fs.foldRight(unit[S, List[A]](List[A]()))((sa, sb) => sa.map2(sb)(_ :: _))
+  }
+
+  def unit[S,A](a: A): State[S,A] = State(s => (a, s))
```

Comment:
`unit` and `sequence` should be defined on a companion object, they don't interact with any of `State`'s field (well, it's only field, `run`).

Reply:
That makes sense.  Is that the only criteria I should be looking for?   I wasn't sure when was appropriate

Reply:
Well, I probably should have said instead that it shouldn't in on the case class because it forces you to have an instance of State to be able to call sequence or unit, which in this case would be a problem. But in other context it can be different.